flang/include/flang/Lower/PFTBuilder.h

```diff
@@ -547,6 +551,8 @@ struct FunctionLikeUnit : public ProgramUnit {
   FunctionLikeUnit(FunctionLikeUnit &&) = default;
   FunctionLikeUnit(const FunctionLikeUnit &) = delete;
 
+  bool isRecursive() { return isMainProgram() ? false : recursiveFunction; }
```

Comment:
Name resolution is already capturing the RECURSIVE/NON_RECURSIVE keywords into the symbol. So this could be:
```
bool isRecursive() {
 if (isMainProgram())
  return false;
  auto sym = getSubprogramSymbol();
  return sym.attrs().test(Fortran::semantics::Attr::RECURSIVE) || 
       (!sym.attrs().test(Fortran::semantics::Attr::NON_RECURSIVE) && defaultRecursiveFunctionSetting()); 
}
```
Then there is no need to re analyze the parse-tree.
Regarding the `non-recursive-procedures` flag, I guess the front-end should deal with this and set the default RECURSIVE, NON_RECURSIVE flag accordingly. I do not think this is the case currently however.

Reply:
Good point on the symbol table. Thanks.

There is no driver, thus no switches, etc.